File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/BaseClient.java
package no.bankaxept.epayment.client.base;

import java.net.URL;
import java.time.Clock;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.ServiceLoader;
import java.util.function.Supplier;
import no.bankaxept.epayment.client.base.accesstoken.AccessTokenProvider;
import no.bankaxept.epayment.client.base.accesstoken.AccessTokenPublisher;
import no.bankaxept.epayment.client.base.accesstoken.ScheduledAccessTokenPublisher;
import no.bankaxept.epayment.client.base.accesstoken.StaticAccessTokenPublisher;
import no.bankaxept.epayment.client.base.accesstoken.SuppliedAccessTokenPublisher;
import no.bankaxept.epayment.client.base.http.HttpClient;
import no.bankaxept.epayment.client.base.http.HttpResponse;
import no.bankaxept.epayment.client.base.spi.HttpClientProvider;
import reactor.core.publisher.Mono;

public class BaseClient {

  private final static Duration tokenTimeout = Duration.ofSeconds(10);

  private final HttpClient httpClient;
  private final AccessTokenPublisher tokenPublisher;

  private BaseClient(HttpClient httpClient, AccessTokenPublisher tokenPublisher) {
    this.httpClient = httpClient;
    this.tokenPublisher = tokenPublisher;
  }

  private Mono<Map<String, List<String>>> filterHeaders(
      Map<String, List<String>> headers,
      String correlationId,
      boolean hasBody
  ) {
    var filteredHeaders = new LinkedHashMap<>(headers);
    if (correlationId != null) {
      filteredHeaders.put("X-Correlation-Id", List.of(correlationId));
    }
    return tokenPublisher.getAccessToken().timeout(tokenTimeout)
            .map(jwt->{
              filteredHeaders.put(
                  "Authorization",
                  List.of("Bearer " + jwt)
              );
              if (hasBody && !headers.containsKey("Content-Type"))
                filteredHeaders.put("Content-Type", List.of("application/json"));
              return filteredHeaders;
            });
  }

  public Mono<HttpResponse> get(
      String uri,
      Map<String, List<String>> headers
  ) {
    return filterHeaders(headers, null, false)
        .flatMap(filteredHeaders-> httpClient.get(uri, filteredHeaders));
  }

  public Mono<HttpResponse> post(
      String uri,
      String body,
      String correlationId
  ) {
    return post(uri, body, correlationId, Map.of());
  }

  public Mono<HttpResponse> post(
      String uri,
      String body,
      String correlationId,
      Map<String, List<String>> headers
  ) {
    return filterHeaders(headers, correlationId, true)
        .flatMap(filteredHeaders-> httpClient.post(uri, body, filteredHeaders));
  }

  public Mono<HttpResponse> delete(
      String uri,
      String correlationId
  ) {
    return filterHeaders(Map.of(), correlationId, false)
        .flatMap(filteredHeaders-> httpClient.delete(uri, filteredHeaders));
  }

  public Mono<HttpResponse> put(
      String uri,
      String body,
      String correlationId
  ) {
    return filterHeaders(Map.of(), correlationId, true)
        .flatMap(filteredHeaders-> httpClient.put(uri, body, filteredHeaders));
  }

  public void shutDown() {
    tokenPublisher.shutDown();
  }

  public static class Builder {

    private final HttpClient httpClient;
    private AccessTokenPublisher tokenPublisher;

    public Builder(URL resourceServerUrl) {
      this.httpClient = ServiceLoader.load(HttpClientProvider.class)
          .findFirst()
          .map(httpClientProvider -> httpClientProvider.create(resourceServerUrl.toString()))
          .orElseThrow();
    }

    public Builder withStaticToken(String token) {
      this.tokenPublisher = new StaticAccessTokenPublisher(token);
      return this;
    }

    public Builder withSuppliedToken(Mono<String> tokenSupplier) {
      this.tokenPublisher = new SuppliedAccessTokenPublisher(tokenSupplier);
      return this;
    }

    public Builder withScheduledToken(URL authorizationServerUrl, String id, String secret) {
      AccessTokenProvider accessTokenProvider = new AccessTokenProvider.Builder()
          .httpClient(httpClient)
          .url(authorizationServerUrl)
          .clientCredentials(id, secret)
          .build();
      this.tokenPublisher = new ScheduledAccessTokenPublisher(accessTokenProvider, Clock.systemDefaultZone());
      return this;
    }

    public Builder withScheduledToken(URL authorizationServerUrl, String id, String secret, Clock clock) {
      AccessTokenProvider accessTokenProvider = new AccessTokenProvider.Builder()
          .httpClient(httpClient)
          .url(authorizationServerUrl)
          .clientCredentials(id, secret)
          .build();
      this.tokenPublisher = new ScheduledAccessTokenPublisher(accessTokenProvider, clock);
      return this;
    }

    public BaseClient build() {
      return new BaseClient(httpClient, tokenPublisher);
    }
  }


}

File: csharp/BaseClient/src/BaseClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Reactive.Linq;
using System.Text;
using System.Threading.Tasks;
using BankAxept.Epayment.Client.Base.AccessToken;

namespace BankAxept.Epayment.Client.Base {

    public class BaseClient  {
        
        private static readonly TimeSpan tokenTimeout = TimeSpan.FromSeconds(10);
        private readonly HttpClient client;
        private readonly IAccessTokenPublisher tokenPublisher;
        
        public BaseClient(HttpClient httpClient, IAccessTokenPublisher tokenPublisher) {
            client = httpClient;
            this.tokenPublisher = tokenPublisher;
        }

        private HttpRequestHeaders FilterHeaders(HttpRequestHeaders headers, string correlationId, bool hasBody) {
            if (correlationId != null) {
                headers.Add("X-Correlation-Id", correlationId);
            }

            headers.Add("Authorization", $"Bearer {new AccessTokenSubscriber(tokenPublisher).Get(tokenTimeout)}");
            return headers;
        }

        public async Task<RequestStatus> Get(string uri) {
            var requestMessage = new HttpRequestMessage (HttpMethod.Get, uri);
            FilterHeaders(requestMessage.Headers, null, false);
            var response = await client.SendAsync(requestMessage);
            var body = await response.Content.ReadAsStringAsync();
            return RequestStatusParser.Parse(response.StatusCode, body);
        }

        public async Task<RequestStatus> Post(string uri, string body, string correlationId) {
            var content = new StringContent(body, Encoding.UTF8, "application/json");
            var requestMessage = new HttpRequestMessage {
                Method = HttpMethod.Post,
                RequestUri = new Uri($"{client.BaseAddress}{uri}"),
                Content = content
            };
            
            FilterHeaders(requestMessage.Headers, correlationId, true);
            var response = await client.SendAsync(requestMessage);
            var responseBody = await response.Content.ReadAsStringAsync();
            return RequestStatusParser.Parse(response.StatusCode, responseBody);
        }

        public async Task<RequestStatus> Put(string uri, string body, string correlationId) {
            var content = new StringContent(body, Encoding.UTF8, "application/json");
            var requestMessage = new HttpRequestMessage {
                Method = HttpMethod.Post,
                RequestUri = new Uri(uri, UriKind.Relative),
                Content = content
            };

            FilterHeaders(requestMessage.Headers, correlationId, true);
            var response = await client.SendAsync(requestMessage);
            var responseBody = await response.Content.ReadAsStringAsync();
            return RequestStatusParser.Parse(response.StatusCode, responseBody);
        }

        public async Task<RequestStatus> Delete(string uri, string correlationId) {
            var requestMessage = new HttpRequestMessage(HttpMethod.Delete, uri);
            FilterHeaders(requestMessage.Headers, correlationId, false);
            var response = await client.SendAsync(requestMessage);
            var body = await response.Content.ReadAsStringAsync();
            return RequestStatusParser.Parse(response.StatusCode, body);
        }

        public class Builder {
            private readonly HttpClient client;
            private IAccessTokenPublisher tokenPublisher;

            public Builder(string resourceServerUrl) {
                var baseAddress = resourceServerUrl.TrimEnd('/');
                client = new HttpClient {
                    BaseAddress = new Uri(baseAddress)
                };
            }

            public Builder WithScheduledToken(string authorizationServerUrl, string clientId, string clientSecret) {
                tokenPublisher =
                    new ScheduledAccessTokenPublisher.Builder(new HttpClient())
                        .Url(new Uri(authorizationServerUrl))
                        .ClientCredentials(clientId, clientSecret)
                        .GrantType("client_credentials")
                        .Build();
                
                return this;
            }

            public BaseClient Build() {
                return new BaseClient(client, tokenPublisher);
            }

            //public Builder WithStaticToken(string token) {
            //    TokenPublisher = new StaticAccessTokenPublisher(token);
            //}

            // public Builder WithScheduledToken(Uri authorizationServerUrl, string id, string secret) {
            //     tokenPublisher = new ScheduledAccessTokenPublisher(client, authorizationServerUrl);
            //     return this;
            // }

            // public BaseClient Build() {
            //     return new BaseClient(Client, TokenPublisher);
            // }            
        }
    }
}